power_grid_model_c/power_grid_model/include/power_grid_model/common/logging_impl.hpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "logging.hpp"
+
+#include <string_view>
+#include <vector>
+
+namespace power_grid_model::common::logging {
+class NoLogger : public Logger {
+    void log(LogEvent /*tag*/) override {}
+    void log(LogEvent /*tag*/, std::string_view /*message*/) override {}
+    void log(LogEvent /*tag*/, double /*value*/) override {}
+    void log(LogEvent /*tag*/, Idx /*value*/) override {}
+    std::unique_ptr<Logger> clone() const override { return std::make_unique<NoLogger>(); }
+};
+
+class LogDispatcher final : public LogDispatch {
+  public:
+    void log(LogEvent tag) override { log_impl(tag); }
+    void log(LogEvent tag, std::string_view message) override { log_impl(tag, message); }
+    void log(LogEvent tag, double value) override { log_impl(tag, value); }
+    void log(LogEvent tag, Idx value) override { log_impl(tag, value); }
+    void registrar(Logger* logger) override {
+        if (logger != nullptr && std::ranges::find(loggers_, logger) == loggers_.end()) {
+            loggers_.push_back(logger);
+        }
+    }
+    void deregistrar(Logger* logger) override {
+        if (logger != nullptr) {
+            std::erase(loggers_, logger);
+        }
+    }
+    std::unique_ptr<Logger> clone() const override { return std::make_unique<LogDispatcher>(); }
+
+    LogDispatcher() = default;
+    LogDispatcher(const LogDispatcher&) = delete;
+    LogDispatcher& operator=(const LogDispatcher&) = delete;
+    LogDispatcher(LogDispatcher&&) = default;
+    LogDispatcher& operator=(LogDispatcher&&) = default;
+    ~LogDispatcher() override = default;
+
+  private:
+    std::vector<Logger*> loggers_;
+
+    template <typename... T>
+    constexpr void log_impl(LogEvent tag, T&&... values)
+        requires requires(Logger log_) {
+            { log_.log(tag, values...) };
+        }
+    {
+        if (loggers_.size() == 1 && loggers_.front() != nullptr) { // allows perfect forwarding to log messages
+            loggers_.front()->log(tag, std::forward<T>(values)...);
+        } else {
+            for (auto& logger : loggers_) {
+                if (logger != nullptr) {
+                    logger->log(tag, values...);
+                }
+            }
+            capturing::into_the_void(std::forward<T>(values)...);
```

Comment:
values can be any count of arguments and maybe not all get used in logging and hence unused ones get sent to void?

Reply:
by C++ guidelines: every object that was moved into the current function shall be moved again (for performance reasons).

Forwarding means that the object could be passed by `const&` or be moved into this object (or captured by value for small objects). You don't know. Therefore, you need to forward objects provided to this function again. But you can only forward once (otherwise, you might be moving an already moved object) and exactly once (cfr. the guideline)

`capturing::into_the_void` is a sink that consumes objects that would otherwise not be moved into this function.